17 decouple conf manager and visor rpc init

fkengun opened the pull request:
I am still using the global ConfigurationManager object CHRONOLOG_CONF. By default, it is initialized with all default values. The ChronoVisor2 takes an object of ConfigurationManager with all values read from a JSON file in its constructor and copies it to the global CHRONOLOG_CONF. rpc-related stuff is initialized inside init(). It is called after the global CHRONOLOG_CONF is set. Thus, when the rpc reads configurations from CHRONOLOG_CONF, everything is set according to the configuration file.

fkengun committed:
Add function to load configurations from a JSON file

fkengun committed:
Remove unused global variable in ChronoVisor

fkengun committed:
Allow passing configuration file to ChronoVisorServer2 constructor to set configuration

fkengun committed:
Loop over all keys in the JSON configuration file to check if supported configurations are available
Change constructor of ChronoLogClient to accept a configuration file as argument
SERVER_LIST in default configuration is initialized to {"localhost"}

fkengun committed:
FIXME: test.sh for isolating cpu cores does not work on Ubuntu 22.04 with cgroup v2

fkengun committed:
Add number of service threads for client in ConfigurationManager
Rename server configurations to visor configurations

fkengun committed:
Fix no_entry error

fkengun committed:
Add a new Visor constructor to accept a ConfigurationManager object to set configurations
Add a new Client constructor to accept a ConfigurationManager object to set configurations
Use configure_file in CMakeLists.txt to make sure updated configuration file is copied during building
Consistent logging using LOGI and LOGE

fkengun committed:
Fix wrong logic in edit_chronicle_attr, it should insert a new record instead of failing if the property does not exist

fkengun committed:
Fix missing return in get_chronicle_attr
Fix building warnings of -Wreturn-type

fkengun committed:
Fix wrong return value in destroy_story
Fix double free error in release mode by rename debug output switch from NDEBUG, which conflicts with the builtin macro, to NODEBUG
Use custom ASSERT macro to have more info in assert error output

fkengun committed:
Fix switch not reachable building error

fkengun committed:
Add two configurations for drift rate calculation sleep control

fkengun committed:
Add two configurations for clocksource

fkengun committed:
Minor fix, no need to have a ConfigurationManager as a member in Visor

fkengun committed:
Add missing json files

Update CMakeLists.txt accordingly

ibrodkin commented on ChronoAPI/ChronoLog/include/ClocksourceManager.h:
```diff
@@ -15,10 +15,10 @@
+    TSC = 2,
+    UNKNOWN = 3
 };
 
```
Do we need an UNKNOIWN case for Clocksource?  Is UNKNOWN used anywhere in the code?  In any case if the enum should include  UNKNOWN let's make it zero, it's a good practice to assign zero value to a zero concept.

ibrodkin commented on ChronoAPI/ChronoLog/include/RPCVisor.h:
```diff
@@ -25,9 +25,9 @@ class RPCVisor {
-                ->GetRPC(CHRONOLOG_CONF->RPC_BASE_SERVER_PORT);
+                ->GetRPC(CHRONOLOG_CONF->RPC_BASE_VISOR_PORT);
         set_prefix("ChronoLog");
-        if (CHRONOLOG_CONF->IS_SERVER) {
```
RPCVisor  does know it's own identity , IS_SERVER or IS_VISOR check is redundant

ibrodkin commented on ChronoAPI/ChronoLog/include/RPCFactory.h:
```diff
@@ -18,11 +18,11 @@ class ChronoLogRPCFactory {
-        CHRONOLOG_CONF->RPC_BASE_SERVER_PORT = temp;
+        CHRONOLOG_CONF->RPC_BASE_VISOR_PORT = temp;
         return rpc;
     }
```
Are we assigning CHRONOLOG_CONF-> RPC_BASE_VISOR_PORT a new value here? I know this code is inherited from hcl, where every RPC instance is both a server and a client , this is not our Visor-Client RPC relationship.  In any case ConfigurationManager contents should not be manipulated by other modules, let's revisit this use case

ibrodkin commented on ChronoAPI/ChronoLog/include/RPCVisor.h:
```diff
@@ -60,7 +62,11 @@ class RPCVisor {
+        } else {
+            return CL_ERR_UNKNOWN;
+        }
     }
```
ClientRegistry, MetadataMaps, etc should be instantiated before any of the RPC communication objects are, no need to check validity of the pointer on every client API call. If we structure our code correctly g_clientRegistryManager == nullptr case would be prevented by the compiler.  Either ClientRegistry::GetInstance() should be called in RPCVisor constructor or the reference to a valid ClientRegistry instance should be passed to RPCVisor constructor.   I think Aparna already started making GetInstance changes on a different branch...

ibrodkin commented on ChronoAPI/ChronoLog/include/ConfigurationManager.h:
```diff
@@ -9,58 +9,110 @@
 #include "log.h"
+#include "errcode.h"
+#include "ClocksourceManager.h"
 
```
ClocksourceManager should be part of ConfigurationManager.h
Configuration Manager is only responsible for reading in the configuration values from the JSON formatted file and doing basic validation on them ( valid ip string , etc)
ConfigurationManager should not be instantiating ClocksourceManager.  ConfigurationManager is a Singleton that all other modules will discover through it's GetInstance() function and pull the configuration values they need  from ConfigurationManager.

fkengun committed:
Add ClocksourceManager to ConfigurationManager

Remove redundant UNKNOWN clocksource
Remove redundant TimeManager
Remove redundant null pointer check in ChronicleMetaDirectory
Remove redundant IS_VISOR check in RPCVisor

ibrodkin reviewed (changes requested):
Could you, please, break the JSON object into sections? Something like this:
{process:"chrono_visor" 
     {clock : { clock_type:"TSC" }{..} }
     {client_rpc : {..all rpc parameters}
     { admin_rpc: { rpc parameters for communication with keepers}
     { authentication : { type and location of authentication module}
     { ...}
}

fkengun committed:
Hierarchical configuration file design

All ChronoLog executables use the same copy of configuration files.
RPC configurations for communications between each pair in Client, Visor and Keeper are separated.

fkengun committed:
Remove ClocksourceManager from ConfigurationManager

Add ClocksourceManager to Visor and Client

fkengun committed:
Move ClocksourceManager.h out of ConfigurationManager.h to ChronoVisorServer2.h and client.h

fkengun committed:
Merge branch 'develop' into 17-decouple-conf-manager-and-visor-rpc-init

# Conflicts:
#	ChronoAPI/ChronoLog/include/RPCVisor.h
#	ChronoAPI/ChronoLog/include/global_var_visor.h
#	ChronoAPI/ChronoLog/include/rpc.h
#	ChronoVisor/include/ChronoVisorServer2.h
#	ChronoVisor/src/ChronicleMetaDirectory.cpp
#	ChronoVisor/src/ChronoVisorServer2.cpp
#	Client/include/RPCClient.h
#	Client/include/client.h
#	Client/src/client.cpp

fkengun committed:
Fix missing brace in rpc.h

Update test case with the new configuration hierarchy

fkengun committed:
Re-merge develop to branch 17

Re-fix wrong brace in rpc.h leads to Visor launch failure

fkengun committed:
Merge remote-tracking branch 'origin/17-decouple-conf-manager-and-visor-rpc-init' into 17-decouple-conf-manager-and-visor-rpc-init

# Conflicts:
#	ChronoAPI/ChronoLog/include/rpc.h
#	test/integration/Client/CMakeLists.txt

ibrodkin reviewed (approved):
please, merge
